File: 71-simplify-path/simplify-path.java
class Solution {
    public String simplifyPath(String path) {
        Deque<String> deq = new ArrayDeque<>();
        for(String dir : path.split("/")) {
            if(dir.isEmpty() || dir.equals(".")) continue;
            else if(dir.equals("..")) {
                if(!deq.isEmpty()) deq.removeFirst();
            }else deq.addFirst(dir);
        }

        StringBuilder result = new StringBuilder();
        while(!deq.isEmpty()) {
            result.append('/').append(deq.removeLast());
        }
        return result.isEmpty() ? "/" : result.toString();
    }
}

File: 71-simplify-path/simplify-path.cpp
class Solution {
public:
    string simplifyPath(string path) {
        deque<string> stack;

        stringstream ss(path);
        string dir;
        while(getline(ss, dir, '/')) {
            if(dir.empty() || dir == ".") continue;
            else if(dir == "..") {
                if(!stack.empty()) stack.pop_back();
            }else stack.push_back(dir);
        }

        string result;
        while(!stack.empty()) {
            result+= "/" + stack.front();
            stack.pop_front();
        }
        return result.empty() ? "/" : result;
    }
};